### PR TITLE
First Pass and Tests

### PR DESCRIPTION
**Pull request by film42**
Somewhat of a WIP on how the system should look. This is inspired by buttercoin's Order models and scala-cucumber-matching-engine's blog post series. This mostly serves as a sketch for now.

### CODE REVIEW

**Comment by film42** [comment]
Ok, so this now includes tests for the following:
- Matching Engine:
  - Limit Order
  - Market Order
- Order Book
  - Buy / Sell but implemented only for Limit and Market Orders

I think this is a good point to merge what's here though.

Can you guys let me know what you think?

cc @marioagarcia @reedcwilson @sgbench @Calmquist @cdbelyeu

**Comment by marioagarcia** [comment]
I think it's ready for merging

**Comment by cdbelyeu** [comment]
I agree. I don't really know the language, though, so I didn't want to
merge it down...

File: src/main/scala/darkpool/book/OrderBook.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  def decreaseTopBy(quantity: Double) {
+    def decreaseTopOfLimitBook(): Unit = limitBook match {
+      case ((bookLevel, orders) :: tail) =>
+        val (top :: rest) = orders
+        limitBook = (quantity == top.quantity, rest.isEmpty) match  {
+          case (true, true) => tail
+          case (true, false) => (bookLevel, rest) :: tail
+          case _ => (bookLevel, top.decreasedBy(quantity) :: rest) :: tail
+        }
+      case _ => throw new IllegalStateException("Bad state trying to decrease top of limit book!")
```
Comment (cdbelyeu) [inline src/main/scala/darkpool/book/OrderBook.scala]: Real FP of you to throw exceptions.

**Pull request merged**